SPARK-3996: Shade Jetty in Spark deliverables.

pwendell opened the pull request:
This patch relocates Jetty classes inside of the assembly jar in Spark.
It does not modify Spark's disclosed dependency on Jetty, since we
already ask users to mark Spark and its dependencies as "provided"
anyways. This is a small fix that hopefully can be used as a template
for other locations in which we shade libraries.

This is a simpler version of #2984. If merged, this closes #2984.

mccheah commented:
Hi,

I was wondering if this is still being updated?

pwendell commented:
Hey All,

I don't see where spark-submit fits into the issue of having a conflicting jetty library? If you have an application that requires a conflicting version of jetty, it doesn't matter one way or the other whether you use spark-submit.

There are certainly applications that do not user spark-submit, and I don't think we have any issue with that. Databricks Cloud is like this, it embeds Spark. I think we'll always support that, it will just be more work for those applications to deal with setting up classpaths, etc in the proper way.

The issue here is just a genuine dependency conflict between a user application and Spark. We don't have a general purpose way of solving those.

pwendell commented:
One case I suppose where it could be relevant is if we only shaded Jetty inside of our assembly jar. That would only help users of spark-submit (also, it would be much easier). But in the past I don't think we did that kind of shading.

mccheah commented:
Looks like we're on the same page. However I believe this still raises the question of how to best do the shading itself. It looks like the short-term solution is to shade Jetty in all of the sub-projects, but that still sounds less than ideal, especially if we're looking for something repeatable and generic..

pwendell committed:
[WIP] [SPARK-3996]: Shade Jetty in Spark deliverables

This patch piggy-back's on @vanzin's work to simplify the Guava shading,
and adds Jetty as a shaded library in Spark. Other than adding Jetty,
it consilidates the <artifactSet>'s into the root pom. I found it was
a bit easier to follow that way, since you don't need to look into
child pom's to find out specific artifact sets included in shading.

pwendell commented:
Closing this to re-open with a new version.
